File: returned/python/today16/course_schedule.py
# https://leetcode.com/problems/course-schedule/description/
from typing import List

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        prereqMap = { i:[] for i in range(numCourses) }
        for course, prereq in prerequisites:
            prereqMap[course].append(prereq)
        
        visited = set()
        def dfs(course: int) -> bool:
            if course in visited:
                return False
            if prereqMap[course] == []:
                return True
            
            visited.add(course)
            for prereq in prereqMap[course]:
                if not dfs(prereq):
                    return False
            visited.remove(course)
            prereqMap[course] = []
            return True

        for course in prereqMap:
            if not dfs(course):
                return False
        return True

File: csharp/Leet/Today4/CourseSchedule.cs
// https://leetcode.com/problems/course-schedule/description/
namespace Today4.CourseSchedule;

public class Solution
{
    public bool CanFinish(int numCourses, int[][] prerequisites)
    {
        var graph = new Dictionary<int, List<int>>();
        var indegree = new Dictionary<int, int>();
        foreach (var prereq in prerequisites)
        {
            if (!graph.ContainsKey(prereq[0]))
                graph[prereq[0]] = new List<int>();
            graph[prereq[0]].Add(prereq[1]);

            if (!indegree.ContainsKey(prereq[1]))
                indegree[prereq[1]] = 1;
            else
                indegree[prereq[1]] += 1;
        }

        var queue = new Queue<int>();
        for (int course = 0; course < numCourses; course++)
        {
            if (!indegree.ContainsKey(course))
                queue.Enqueue(course);
        }

        var count = 0;
        while (queue.Count > 0)
        {
            var course = queue.Dequeue();
            count++;
            if (graph.ContainsKey(course))
            {
                foreach (var adj in graph[course])
                {
                    indegree[adj] -= 1;
                    if (indegree[adj] == 0)
                        queue.Enqueue(adj);
                }
            }
        }

        return count == numCourses;
    }
}